File: GGAndroid_2.0/GGCore/GGRxJava/src/main/java/io/ganguo/rx/RxHelper.java
package io.ganguo.rx;

import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import io.reactivex.Maybe;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import io.reactivex.functions.Cancellable;
import io.reactivex.subjects.Subject;

/**
 * Created by Roger on 21/07/2017.
 */

public class RxHelper {
    private RxHelper() {
        throw new AssertionError("No instances.");
    }

    public static <T> void checkNull(@Nullable T value, @NonNull String name) {
        if (value == null) {
            throw new NullPointerException(name + " must not be null.");
        }
    }

    public static <T> Maybe<T> createInitialMaybe(@Nullable T initialValue) {
        checkNull(initialValue, "initialValue");
        return Maybe.just(initialValue);
    }

    public static <T> void safeComplete(@NonNull Subject<T> emitter) {
        if (!emitter.hasThrowable() && !emitter.hasComplete()) {
            emitter.onComplete();
        }
    }

    public static void safeDispose(@Nullable Disposable disposable) {
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    public static void safeCancel(@Nullable Cancellable cancellable) {
        if (cancellable == null) {
            return;
        }

        try {
            cancellable.cancel();
        } catch (Exception e) {
            // Ignore the exception.
        }
    }

    public static boolean checkMainThread(Observer<?> observer) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            observer.onSubscribe(Disposables.empty());
            observer.onError(new IllegalStateException(
                    "Expected to be called on the main thread but was " + Thread.currentThread().getName()));
            return false;
        }
        return true;
    }
}

File: GGAndroid_3.0/core/support/rxjava/src/main/java/io/ganguo/rx/utils/RxHelper.kt
package io.ganguo.rx.utils

import io.reactivex.Maybe
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Cancellable
import io.reactivex.subjects.Subject

/**
 * <pre>
 *     author : leo
 *     time   : 2020/05/12
 *     desc   : RxJava Common functions
 * </pre>
 */

/**
 * check value is null
 * @param value T?
 * @param name String
 * @return Unit
 */
fun <T> checkNull(value: T?, name: String): Unit {
    if (value == null) {
        throw NullPointerException("$name must not be null.")
    }
}

/**
 * create MaybeSource
 * @param initialValue T?
 * @return Maybe<T>?
 */
fun <T> createInitialMaybe(initialValue: T?): Maybe<T>? {
    checkNull(initialValue, "initialValue")
    return Maybe.just(initialValue)
}

/**
 * It is safe to carry out onComplete on Subject
 * @param emitter Subject<T>
 */
fun <T> safeComplete(emitter: Subject<T>) {
    if (!emitter.hasThrowable() && !emitter.hasComplete()) {
        emitter.onComplete()
    }
}

/**
 * It is safe to carry out dispose on Disposable
 * @param disposable Disposable?
 */
fun safeDispose(disposable: Disposable?) {
    if (disposable != null && !disposable.isDisposed) {
        disposable.dispose()
    }
}

/**
 * It is safe to carry out cancel on Cancellable
 * @param cancellable Cancellable?
 */
fun safeCancel(cancellable: Cancellable?) {
    if (cancellable == null) {
        return
    }
    try {
        cancellable.cancel()
    } catch (e: Exception) {
        e.printStackTrace()
    }
}